src/igraph_eulerian.c

```diff
@@ -0,0 +1,502 @@
+0 -> if graph is not Eulerian
+1 -> if graph has an euler circuit
+2 -> if graph has an euler path */
+igraph_integer_t is_eulerian_undirected(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+
+    igraph_bool_t res;
+    igraph_inclist_t il;
+    igraph_vector_int_t *incedges;
+    igraph_integer_t odd, i;
+
+    odd = 0;
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res, IGRAPH_WEAK));
+    if (res == 0) {
+        *has_path = 0;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_ALL));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        incedges = igraph_inclist_get(&il, i);
+        if ( igraph_vector_int_size(incedges) % 2 == 1) odd++;
+    }
+
+    igraph_inclist_destroy(&il);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    if (odd > 2) {
+        *has_path = 0;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    /* if pdd count is 2, then semi-eulerian
+    if odd count is 0, then eulerian
+    odd count never is 1 for undirected graph
+    */
+
+    if (odd == 2) {
+        *has_path = 1;
+        *has_cycle = 0;
+    } else {
+        *has_path = 0;
+        *has_cycle = 1;
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+igraph_integer_t is_eulerian_directed(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_es_t incoming, outgoing;
+    igraph_bool_t res_strong, res_weak;
+    igraph_integer_t incoming_excess, incoming_count, outgoing_excess, outgoing_count;
+    igraph_integer_t i;
+
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        IGRAPH_CHECK(igraph_es_incident(&incoming, i, IGRAPH_IN));
+        IGRAPH_CHECK(igraph_es_incident(&outgoing, i, IGRAPH_OUT));
+        
+        IGRAPH_CHECK(igraph_es_size(graph, &incoming, &incoming_count));
+        IGRAPH_CHECK(igraph_es_size(graph, &outgoing, &outgoing_count));
+
+        if (incoming_count != outgoing_count) {
+            if ((incoming_count + 1 == outgoing_count) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+            } else if ((outgoing_count + 1 == incoming_count) && (incoming_excess < 1 && outgoing_excess < 2)) {
+                incoming_excess++;
+            } else {
+                *has_path = 0;
+                *has_cycle = 0;
+                return IGRAPH_SUCCESS;
+            }
+        }
+    }
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_strong, IGRAPH_STRONG));
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_weak, IGRAPH_WEAK));
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong == 1)) {
+        *has_path = 0;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && 
+        (res_strong == 1 || res_weak == 1)) {
```

Comment:
This is incorrect. Weak connectivity is insufficient. Also, `res_strong || res_weak` is the same as `res_weak`.

Reply:
This changed quite a bit since. Plus, you don't need strong connectivity for a Eulerian path (though you do for a Eulerian cycle): the example is the graph 0->1 which is not strongly connected (since there's no reaching 0 from 1) but has a Eulerian path (namely, the only edge).

I clarified this with some comments now